dali/pipeline/operators/reader/loader/video_loader.h

```diff
@@ -0,0 +1,219 @@
+  int frame_count_;
+
+  int vid_stream_idx_;
+  int last_frame_;
+
+#ifdef HAVE_AVBSFCONTEXT
+  av_unique_ptr<AVBSFContext> bsf_ctx_;
+#else
+  using bsf_ptr = std::unique_ptr<AVBitStreamFilterContext, BSFDeleter>;
+  bsf_ptr bsf_ctx_;
+  AVCodecContext* codec;
+#endif
+  av_unique_ptr<AVFormatContext> fmt_ctx_;
+};
+
+/// Provides statistics, see VideoLoader::get_stats() and VideoLoader::reset_stats()
+struct VideoLoaderStats {
+    /** Total number of bytes read from disk */
+    uint64_t bytes_read;
+
+    /** Number of compressed packets read from disk */
+    uint64_t packets_read;
+
+    /** Total number of bytes sent to NVDEC for decoding, can be
+     *  different from bytes_read when seeking is a bit off or if
+     *  there are extra streams in the video file. */
+    uint64_t bytes_decoded;
+
+    /** Total number of packets sent to NVDEC for decoding, see bytes_decoded */
+    uint64_t packets_decoded;
+
+    /** Total number of frames actually used. This is usually less
+     *  than packets_decoded because decoding must happen key frame to
+     *  key frame and output sequences often span key frame sequences,
+     *  requiring more frames to be decoded than are actually used in
+     *  the output. */
+    uint64_t frames_used;
+};
+
+
+class VideoLoader : public Loader<GPUBackend, SequenceWrapper> {
+ public:
+  explicit inline VideoLoader(const OpSpec& spec,
+    const std::vector<std::string>& filenames)
+    : Loader<GPUBackend, SequenceWrapper>(spec),
+      count_(spec.GetArgument<int>("count")),
+      output_height_(spec.GetArgument<int>("height")),
+      output_width_(spec.GetArgument<int>("width")),
+      height_(0),
+      width_(0),
+      filenames_(filenames),
+      // TODO(spanev) handle device_id != 0
+      device_id_(0),
+      codec_id_(0),
+      done_(false) {
+  }
+
+  void init() {
+
+    /* Required to use libavformat: Initialize libavformat and register all
+     * the muxers, demuxers and protocols.
+     */
+    av_register_all();
+
+    // TODO(spanev) Implem several files handling
+    total_frame_count_ = get_or_open_file(filenames_[0]).frame_count_;
+
+    // TODO(spanev) to change after deciding what we want
+    int seq_per_epoch = total_frame_count_ / count_;
+    frame_starts_.resize(seq_per_epoch);
+    for (int i = 0; i < seq_per_epoch; ++i) {
+      frame_starts_[i] = i * count_;
+    }
+    // TODO(spanev) change random engine
+    auto rng = std::default_random_engine {};
+    std::shuffle(std::begin(frame_starts_), std::end(frame_starts_), rng);
+    current_frame_idx_ = 0;
+
+    thread_file_reader_ = std::thread{&VideoLoader::read_file, this};
+  }
+
+  ~VideoLoader() noexcept {
+    done_ = true;
+    send_queue_.cancel_pops();
+    if (vid_decoder_) {
+      vid_decoder_->finish();
+    }
+    if (thread_file_reader_.joinable()) {
+      try {
+        thread_file_reader_.join();
+      } catch (const std::system_error& e) {
+        std::cerr << "System error joining thread: "
```

Comment:
Do we really want to recover from such an error with just a message to the console?

Reply:
This is completely a WIP and I'm still working of the general architecture of the Operator, which error handling is part of